DEV-43744-alert-eval-api

yasmin-tr opened the pull request:
**What is this feature?**

Logzio Alert Evaluation API

* contains basic functionality for alert evaluation api
* including headers and necessary files and models to support evaluation on logzio datasources

**Explaining the main logic and changes:** 
We added a new internal api for evaluating alerts externally from the alert-manager.
External in this meaning is that the evaluation is called outside the alert-manager and not as part of the engine schedule as it is today. 
To be able to reuse the evaluation code as much as possible, we are using the ScheduleService code to run alert evaluation. 
In order to use ScheduleService, but without having it running the evaluation, I needed to add a configuration not to execute the evaluations from the scheduler. That way we have all the management of the rule routines and updates, but only the execution is disabled, and then can run from our API.
For running the external evaluation I also added another function for the Scheduler interface that the API is using.

External notification support -
In order to send the notification in an external API, added a LogzioRouter that implements the Send interface which sends the PostableAlerts to the url coming from configuration

Logzio datasources support required the following changes:
* adding headers that contains the user-context passed by the API
* propagating the headers into the evaluation request in relevant clients - elasticsearch and prometheus
* adding url-override from the evaluation request passed by the API
* overriding the url of the datasource for the relevant one so it will call the logzio APIs for query

**Why do we need this feature?**

metrics alerts evaluation process

**Who is this feature for?**

Logzio Metrics Alerts


**Which issue(s) does this PR fix?**:

DEV-43744
DEV-43883
DEV-43889

Added code from the following tickets:
* DEV-31493 Override datasource URL and pass custom headers to alert rule evaluator (partly)
* DEV-17927 - add LogzIoHeaders, Add error msg, change time to StartTime 
* DEV-30169: api alert evaluation
* DEV-16492 Support external alert evaluation 
* (no ticket) Upgrade to 8.4.0 
* DEV-18005 - add requestId in error message
* DEV-20400 - Grafana alerts evaluation - set 'accountsToSearch' query param

yasmin-tr committed:
Added code from tickets:

*DEV-31493 Override datasource URL and pass custom headers to alert rule evaluator
*DEV-17927 - add LogzIoHeaders, Add error msg, change time to StartTime
*DEV-30169, DEV-30170: api alert evaluation and alert processing
(processing in comments)
*DEV-16492 Support external alert evaluation
*(no ticket) Upgrade to 8.4.0
*DEV-18900 - fix For check to be inclusive

Jonathan-Eng reviewed (commented):
as discussed. the pr is not ready yet for review

yasmin-tr committed:
added error_response.logzio.go
DEV-18005 - add requestId in error message

yasmin-tr committed:
Switched to new jira tickets:
DEV-43744 , DEV-43883 , DEV-43895

Removed all code changes related to old alerting which isn't used, therefor not needed.

yasmin-tr committed:
Change implementation to use the routine in scheduler instead of copy-pasting thing relevant code into our API.

!!still work in progress!!

need to:
*remove commented code
*add logz headers somehow to request
*fix scheduler so it will not evaluate so we can use it for the api
*remove processing part

yasmin-tr committed:
DEV-43744 - Changes to evaluation API
* not returning results in API and keeping the processing part in eval
* added a logzio router to replace the regular router and send notifications to logz internal api (in config)
* removed unused code of processing
* changed the evaluation api to run as bulk

yasmin-tr committed:
Fix compilation errors

yasmin-tr committed:
removed unused code

ohadza commented on pkg/services/ngalert/api/api.go:
```diff
@@ -2,6 +2,7 @@ package api
 
 import (
 	"context"
+	"github.com/grafana/grafana/pkg/services/ngalert/schedule"
```
mark with `// LOGZ.IO GRAFANA CHANGE`

ohadza commented on pkg/services/ngalert/api/api.go:
```diff
@@ -75,6 +76,7 @@ type API struct {
 	Tracer               tracing.Tracer
 	AppUrl               *url.URL
 	UpgradeService       migration.UpgradeService
+	Schedule             schedule.ScheduleService
```
mark with `// LOGZ.IO GRAFANA CHANGE`

ohadza commented on pkg/services/ngalert/api/api_logzio_alerting.go:
```diff
@@ -0,0 +1,62 @@
+}
+
+// NewLogzioAlertingApi creates a new LogzioAlertingApi instance
+func NewLogzioAlertingApi(service *LogzioAlertingService) *LogzioAlertingApi {
```
Didn't we say its redundant? (leave only the service)

yasmin-tr replied:
this is the structure of all APIs in this project. there is a class of "XXXApi" that has the route to the endpoint, and the function to call the service class that contains the main logic.
i think what we said is that maybe the LogzioAlertingService can be redundant and we can put the logic directly in this class. but actually i think this structure has some sense in it.. also i'm pretty sure we'll want it also for later when more logic/apis will be added (or at least it gives the infra for it).

ohadza commented on pkg/services/ngalert/api/tooling/definitions/alertmanager.go:
```diff
@@ -4,6 +4,9 @@ import (
 	"fmt"
+
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+	// LOGZ.IO GRAFANA CHANGE :: End
```
where is the start :)

yasmin-tr replied:
probably some merge messed this up 😅  will fix it 👍

ohadza commented on pkg/services/ngalert/schedule/schedule.go:
```diff
@@ -467,6 +551,8 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
 			))
 		}
+
+		logger.Debug("RuleRoutine evaluation: processing evaluation results")
```
you should cleanup eventually all the logger.Debug(...) and comments ....

yasmin-tr commented on pkg/api/alerting.go:
```diff
@@ -9,6 +9,7 @@ import (
 
 	"github.com/grafana/grafana/pkg/api/dtos"
 	"github.com/grafana/grafana/pkg/api/response"
+	"github.com/grafana/grafana/pkg/models" // LOGZ.IO GRAFANA CHANGE :: DEV-17927 - add LogzIoHeaders
```
done

yasmin-tr commented on pkg/api/alerting.go:
```diff
@@ -179,7 +180,8 @@ func (hs *HTTPServer) AlertTest(c *contextmodel.ReqContext) response.Response {
 
-	res, err := hs.AlertEngine.AlertTest(c.SignedInUser.GetOrgID(), dto.Dashboard, dto.PanelId, c.SignedInUser)
+	// LOGZ.IO GRAFANA CHANGE :: DEV-17927 - add LogzIoHeaders
+	res, err := hs.AlertEngine.AlertTest(c.SignedInUser.GetOrgID(), dto.Dashboard, dto.PanelId, c.SignedInUser, &models.LogzIoHeaders{RequestHeaders: c.Req.Header})
```
it is technically part of the logzio datasources support, but because a lot of the implementation for the logz evaluation includes this i put it in this PR.

i don't know if at that point it's worth the time of taking out of this PR for the next , because the next PR anyway will be based on that and making these changes can add up a day of work i don't see any value for it 😅 
in hindsight i would make the split if i knew about it from the beginning 😓

yasmin-tr committed:
DEV-43889 - Added datasources support for logzio

yasmin-tr commented on pkg/promlib/querydata/request.go:
```diff
@@ -110,27 +110,27 @@ func (s *QueryData) Execute(ctx context.Context, req *backend.QueryDataRequest)
 
 	logger := s.log.FromContext(traceCtx)
-	logger.Debug("Sending query", "start", q.Start, "end", q.End, "step", q.Step, "query", q.Expr)
+	logger.Debug("fetching query", "query", q, "headers", headers)
```
will fix this to be the original - for some reason didn't have the log so i added it , not sure how this happened

yasmin-tr replied:
will fix this to be the original - for some reason didn't have the log so i added it , not sure how this happened

yasmin-tr commented on pkg/promlib/client/client_test.go:
```diff
@@ -94,7 +94,7 @@ func TestClient(t *testing.T) {
 				Step:       1 * time.Second,
 			}
-			res, err := client.QueryRange(context.Background(), req)
+			res, err := client.QueryRange(context.Background(), req, map[string]string{})
```
will add comment

yasmin-tr committed:
Merge remote-tracking branch 'origin/v10.4.x-logzio' into DEV-43744-alert-eval-api

# Conflicts:
#	pkg/services/ngalert/api/api.go
#	pkg/services/ngalert/schedule/schedule.go
#	pkg/setting/setting_unified_alerting.go

yasmin-tr committed:
Merge branch 'v10.4.x-logzio' into DEV-43744-alert-eval-api

yasmin-tr committed:
* Fixes from merge
* Renamed logzio files to same standard
* Changed request object for logzio router notification API and changed default url behaviour

yasmin-tr committed:
* small fixes (comments , removed redundant code)
